File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res, str_list = 0, []
        
        for i in s:
            if i in str_list:
                str_list = str_list[str_list.index(i)+1:]
            str_list.append(i)
            res = max(res,len(str_list))
            
        return res
    
class Solution2:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen = {}
        ret = 0
        p = -1
        for i, c in enumerate(s):
            if c in seen and seen[c] > p:
                p = seen[c]
            seen[c] = i
            ret = max(ret, i - p)
        return ret

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.go
func lengthOfLongestSubstring(s string) int {
	set := map[byte]bool{}
	l := 0
	res := 0

	for r := 0; r < len(s); r++ {
		for set[s[r]] {
			delete(set, s[l])
			l++
		}
		set[s[r]] = true
		res = max(res, r-l+1)
	}
	return res
}